src/monte_carlo.c

```diff
@@ -0,0 +1,240 @@
+    }
+  }
+}
+
+/**
+ * This method executes a complete random playout of actions.
+ * @param node The root node from which the episode is played out.
+ * @param player The current player.
+ * @param game The definition of the game played, the environment in which the
+ * the machine learns.
+ * @return The result of the random playout.
+ */
+int mcEpisode(Node* node, int initialPlayer, Game* game, Board* board) {
+  int score = game->getScore(board, node->action.player);
+  if (score > 1) {
+    if (node->action.player == initialPlayer) {
+      return score;
+    } else {
+      return game->lossValue;
+    }
+  }
+
+  int varPlayer = -node->action.player;
+
+  // Deep copy of board
+  Board simulationBoard;
+  simulationBoard.values = malloc(game->getBoardSize() * sizeof(int8_t));
+  for (int i = 0; i < game->getBoardSize(); ++i) {
+    simulationBoard.values[i] = board->values[i];
+  }
+  simulationBoard.nPlayers = board->nPlayers;
+
+  int nPossibleActions = game->getNumPossibleActions(simulationBoard);
+  Action possibleActions[nPossibleActions];
+  game->getPossibleActions(simulationBoard, possibleActions);
+
+  // Random playout
+  while (nPossibleActions > 0) {
+    // Pick random action
+    int randomActionIdx =
+        linear_congruential_random_generator() * nPossibleActions;
+    if (game->isValidAction(board, &possibleActions[randomActionIdx],
+                            varPlayer)) {
+      // Play action
+      game->playAction(&simulationBoard, &possibleActions[randomActionIdx]);
+
+      // Remove action from possible actions
+      nPossibleActions -= simulationBoard.nPlayers;
+      game->removeAction(randomActionIdx, possibleActions, nPossibleActions);
+      int score = game->getScore(&simulationBoard, varPlayer);
+      if (score > game->drawValue) {
+        free(simulationBoard.values);
+        if (varPlayer == initialPlayer) {
+          return score;
+        } else {
+          return game->lossValue;
+        }
+      }
+      varPlayer = -(varPlayer);
+    }
+  }
+  free(simulationBoard.values);
+  return game->drawValue;
+}
+
+/**
+ * This method backpropagates the result of the playout to the root of the
+ * action tree.
+ * @param node The leaf node from which the backpropagation begins.
+ * @param score The score to backpropagate.
+ */
+void backpropagate(Node* node, int score) {
+  node->nVisits++;
+  node->score += score;
+
+  Node* currentChild = node;
+  Node* currentNode = node->parent;
+  while (currentNode) {
+    currentNode->nVisits += 1;
+    currentNode->score += score;
+    currentChild = currentNode;
+    currentNode = currentChild->parent;
+  }
+}
+
+/**
+ * This method frees the memory used during the Monte Carlo algorithm.
+ * @param node The root node from which the function starts freeing the memory.
+ */
+void freeMCTree(Node* node) {
+  for (unsigned i = 0; i < node->nChildren; ++i) {
+    freeMCTree(&node->children[i]);
+  }
+  free(node->children);
+  node->children = NULL;
```

Comment:
Do you need to set it to NULL after it has been freed

Reply:
To avoid dangling pointers, you think I shouldn't set it to null?